### DISCUSSION TITLE
Configuration to allow auto deletion of Sentry & Saved clips that were already archived and are older than X days.

### DISCUSSION THREAD

**Discussion by thesourcerer** [marcone/teslausb#859] [Ideas]
That would be great to max the disk space for recent clips going back longer when using pi zero and it's not applicable to archive recent clips.

Any tips on how to contribute such feature?

**Comment by Arthur-Philip-Dent** [comment]
Isn’t that anyway already happening?

**Reply by thesourcerer** [reply]
Not that I have noticed. The car used to delete Sentry automatically but now it thinks that space is never running out" so it never does.

**Reply by marcone** [reply]
> Not that I have noticed. The car used to delete Sentry automatically but now it thinks that space is never running out" so it never does.

That has nothing to do with TeslaUSB. The car just sees a USB drive of whatever size you specified in the TeslaUSB config file, it doesn't know anything about what TeslaUSB does behind the scenes or how big the actual storage is.
Over the years, the car firmware has alternated between automatically deleting Sentry events when space is needed and not deleting them and just informing the user that the drive is getting full.

TeslaUSB will delete older Sentry events after archiving.

**Reply by thesourcerer** [reply]
But teslausb makes sure to leave 10gb free for that host at all times isnt it?
With the same car firmware when i put the regular tumbdrive sentry gets auto deleted by the car.

Since i moved to teslausb sentry is never deleted any more and i get 3 weeks of sentry recordings. While all of them were archived to my onedrive, the are still present. Do i need to configure something so teslausb will auto delete them after archive? For me it was deleted only after i deleted them from the car screen, and waited for the next low storage event....

**Reply by marcone** [reply]
> But teslausb makes sure to leave 10gb free for that host at all times isnt it?

No, after archiving TeslaUSB removes files from the "CAM" drive until there is at least 20 GB of free space.

> With the same car firmware when i put the regular tumbdrive sentry gets auto deleted by the car.

Then it should be deleted as needed when using TeslaUSB as well. From the car's point of view TeslaUSB is just  a USB drive.
 
> Since i moved to teslausb sentry is never deleted any more and i get 3 weeks of sentry recordings.

Where are you seeing these sentry recordings? If you're looking at the web interface, then that is expected. The files you see in the web interface are not what the car sees.

**Comment by thesourcerer** [comment]
Not sure I understand.
From the car's perspective, it sees a drive that never runs out, as once it's getting near capacity it frees from space isn't it ?
So if it always sees enough space left, as teslausb clears some before the car has a chance to, it will never initiate deletion on its own.
What am I missing here?
I understand that the files in the web interface are "locally archived" (also, they were ALL archived to onedrive, I checked) and the car should see only a subset of it.
But the only files that work "as expected" is the "Recent clips". while the car keeps only 1 hour regardless of space, the local archive contains much more, and old ones gets deleted several times a day when space is nearing capacity.

For dashcam, the car never deletes anything unless I press the delete button on the screen, thus both the car sees them and the local archive (even after they were uploaded to onedrive). once I delete them from the screen, the car doesn't see them any more, in the next trimming session that comes after some time, they are removed from local archive as well but not before, even though they were synced to onedrive days ago.

For sentry, it happens the same, while I expected the car to delete old ones and then to see the local archive getting trimmed as well, it never happens.
I suspect that this is because the car's trimming never kicks in as teslausb always frees up space from recent clips before it has a chance to hit the capacity threshold, and as long as the car view has a link to the file it cannot be trimmed in the current implementation.

Effectively what happens is that if I take no action in the screen to manually delete dashcam and sentry clips, the local archive for recent clips get smaller and smaller every day.

What I was suggesting is to add 2 configurations:
1. Bool config to say if to trim sentry and dashcam clips after cloud archive
2.  int that says how many days/hours to keep clips locally even though they were archived to cloud (I guess most people would prefer to have some locally in the car as well).

This way, I can set "true" and something like "4 days", and once I park at home, sentry and dashcam get archived, and 
clips that were archived and also 4 days or more older, are removed both from the local archive and the car's view.

**Reply by marcone** [reply]
> Not sure I understand. From the car's perspective, it sees a drive that never runs out,

What do you mean by "never runs out"? The car will see a drive of the size you specified in the TeslaUSB config file. The car is not seeing an infinitely sized drive, and won't see files just disappear from the drive while the drive is connected to the car.

> as once it's getting near capacity it frees from space isn't it ?

No, the drive that the car sees is only cleaned up after archiving, and is not connected to the car while being cleaned. So for example, if the drive is 30 GB in size (the default setting in the config file) and has 15 GB of recordings on it, then what the car will see is:
1. a 30 GB drive that has 15 GB of files and 15 GB of free space
1. that drive being unplugged from the car
1. a few seconds pass
1. a drive being plugged in to the car that has 10 GB of files on it and 20 GB of free space

Again, this only happens after archiving, and the drive is not connected to the car while TeslaUSB removes files. While driving  or parked with Sentry mode enabled, the car is managing the space in the drive.

> So if it always sees enough space left, as teslausb clears some before the car has a chance to, it will never initiate deletion on its own. What am I missing here? I understand that the files in the web interface are "locally archived" (also, they were ALL archived to onedrive, I checked) and the car should see only a subset of it. But the only files that work "as expected" is the "Recent clips". while the car keeps only 1 hour regardless of space, the local archive contains much more, and old ones gets deleted several times a day when space is nearing capacity.
> 
> For dashcam, the car never deletes anything unless I press the delete button on the screen, thus both the car sees them and the local archive (even after they were uploaded to onedrive). once I delete them from the screen, the car doesn't see them any more, in the next trimming session that comes after some time, they are removed from local archive as well but not before, even though they were synced to onedrive days ago.
> 
> For sentry, it happens the same, while I expected the car to delete old ones and then to see the local archive getting trimmed as well, it never happens. I suspect that this is because the car's trimming never kicks in as teslausb always frees up space from recent clips before it has a chance to hit the capacity threshold, and as long as the car view has a link to the file it cannot be trimmed in the current implementation.
> 
> Effectively what happens is that if I take no action in the screen to manually delete dashcam and sentry clips, the local archive for recent clips get smaller and smaller every day.

I don't understand what you mean by "the local archive for recent clips". The only reason I can think of for ever having to manually delete files is if you accumulate lots of Sentry or Saved files, but they're not archived (e.g. if you're parked in a busy area away from home for an extended period of time)
 
> What I was suggesting is to add 2 configurations:
> 
> 1. Bool config to say if to trim sentry and dashcam clips after cloud archive
> 2. int that says how many days/hours to keep clips locally even though they were archived to cloud (I guess most people would prefer to have some locally in the car as well).
> 
> This way, I can set "true" and something like "4 days", and once I park at home, sentry and dashcam get archived, and clips that were archived and also 4 days or more older, are removed both from the local archive and the car's view.

I still don't understand what you think you'll gain by this.

**Reply by thesourcerer** [reply]
What you describe is the behavior that I indeed expect. but it's not what I see happening. 
I have a 256GB SD card with those settings:
export CAM_SIZE=98%
export MUSIC_SIZE=1%
export BOOMBOX_SIZE=1%

Effectively, I see all Sentry and Dashcam clips archived both locally and uploaded to OneDrive,
(Happens several times a day when I park at home near WIFI).
But when I look at the dashcam app in the car, I can still see the archived clips there, with nothing that gets deleted from the car's perspective.

I do see the drive "unplugged" a few seconds after I park (Red X on the dashcam app).

**Reply by marcone** [reply]
> What you describe is the behavior that I indeed expect. but it's not what I see happening.
> I have a 256GB SD card with those settings:
> export CAM_SIZE=98%
> export MUSIC_SIZE=1%
> export BOOMBOX_SIZE=1%

OK, I understand now. You changed the default CAM_SIZE setting without fully understanding the consequences.
I suggest you use fixed sizes instead of percentages (I will be removing the percentage-based settings soon, because many people misunderstand them), and use the default `30G` for CAM_SIZE, unless you have a good reason to change it.

**Reply by thesourcerer** [reply]
OK, good to know it's something I did!
Where can I learn more about how it works ?
so if I have a 256 GB MicroSD, let's say that for music and boombox I only need 1 GB each, 
what is the best value for CAM_SIZE ? is it still 30GB regardless the MicroSD size ?
What happens to the rest of the space? it's allocated for archiving?

My goal eventually is to keep locally on the screen 1-3 days of Sentry/dashcam, upload the rest to the cloud (the car is expected to connect to WIFI at least once a day, usually more) and keep a local archive of recent clips for as much days as possible.
(I am not uploading recent clips to the cloud, as the Pi  w2 wifi is not fast enough to keep up the upload) but I do want them to be on teslausb as locally as much as possible to  be able to pull them from the web interface when needed.

given the above, what settings would you recommend for CAM_SIZE?

**Reply by marcone** [reply]
30 GB seems fine. The remaining space will be used for snapshots, or what you called the "local archive" in your earlier comments.

**Reply by thesourcerer** [reply]
I changed it to 30GB. had to delete backing files , snapshots and /mutable/TeslaCam, then ran the setup script for it to take effect.
Will see how it works now.
What should I expect from snapshots?
files from both recent, sentry and dashcam will be removed by oldest time after uploaded to cloud ?

**Reply by marcone** [reply]
A snapshot is essentially a copy of the "cam" disk image at a certain point in time. By default TeslaUSB will take a snapshot about every hour, so that it can keep a continuous record of RecentClips. Older snapshots are deleted to make room for new ones.

**Reply by thesourcerer** [reply]
Updating thay since I reduced the cam size to 30g all is resolved, videos are removed as expected and as you described. Thanks!
